xla/service/cpu/onednn_matmul_rewriter.cc

```diff
@@ -56,13 +56,41 @@ inline Status ValidateDotDimensionNumbers(
   return OkStatus();
 }
 
+// Whether the element type of instr is compatible with oneDNN kernels.
+// TODO(intel-tf): Restict compatible types based on instruction kind.
+inline bool CompatibleElementType(const HloInstruction* instr) {
+  PrimitiveType element_type = instr->shape().element_type();
+  return element_type == BF16 || element_type == F32 || element_type == F16;
+}
+
+// Type conversion from and to any of BF16 and FP32.
+// TODO(intel-tf): Support more types when enabled.
+template <typename Pattern>
+inline auto SupportedConvert(Pattern pattern) {
+  auto supported_convert = [](const HloInstruction* instr) -> bool {
+    return CompatibleElementType(instr) &&
+           CompatibleElementType(instr->operand(0));
+  };
+  return m::Convert(pattern).WithPredicate(supported_convert);
+}
+
+template <typename Pattern>
+inline auto SupportedConvert(HloInstruction** convert, Pattern pattern) {
```

Comment:
@penpornk Thanks for the suggestions. Since the patterns are written as templatized recursive structures and the return type of the above function is based on type deduction, it cannot handle conditionals (`pattern == std::nullopt ? <expr_1> : <expr_2>`) in the return statement. In other words, return type deduction could not be statically determined.

Reply:
Thank you for the clarification!